### ISSUE TITLE
Depletes inotify limit on remote session

### ISSUE THREAD

**Issue by maciejmatczak** [microsoft/pylance-release#3001]
## Environment data

-   Language Server version: 2022.6.30
-   OS and version: Windows 10 with remote on CentOS 7.3
-   Python version (& distribution if applicable, e.g. Anaconda): 3.8, custom build

## Repro Steps

1. Connect to a remote repository with one python file, open a file
2. Wait
3. VS Code complaints about not being able to watch the repository due to a limit

## Expected behavior

Pylance should not try to watch the lib folder, site-packages (?). 

## Actual behavior

Pylance watches lib folders, in my case, it's 100k files, in @phord's case it's 1.2kk (https://github.com/microsoft/pylance-release/issues/2914#issuecomment-1166390438)

## Logs

All of my comments with that one being the most recent https://github.com/microsoft/vscode-python/issues/19234#issuecomment-1175102144

**Comment by maciejmatczak** [comment]
I have no idea what this new (aggressive?) file watch strategy solves, but it is not achieved without a drawback.

There are multiple issues in Pylance, Python extension, VS Code and Remote extension created, as this one caused a bit of a chaos, it seems.

For sure, this issue was not a thing month/two ago.

**Comment by heejaechang** [comment]
@maciejmatczak it is due to a change in our side to use file watcher provided by host (vscode) so that pylance can be used in multiple hosts (vs, vscode, azure and etc)

**Comment by maciejmatczak** [comment]
I would appreciate if we could leave that one open (and maybe count upvotes?) in a "blocked by" manner (blocked by https://github.com/microsoft/vscode/issues/137872, as I understand).

While I understand the gains (and appreciate sharing the reasoning, thanks!), it seems that this decision did have some negative impact for some.

**Comment by heejaechang** [comment]
I tagged "need decision" to see whether we should provide hidden option to turn off library watching until root issue is fixed by vscode. we had many issues opened by users since we didn't support library watching as well, so we can't just stop watching due to the limitation on linux.

**Comment by binh-vu** [comment]
Can you please shed some light on the paths that Pylance send to vscode asking to watch look like?

I takes a look at my `sys.path` and found all of them are absolute paths. People from vscode team said that full absolute path exclude are supported on linux ([link](https://github.com/microsoft/vscode/issues/3998#issuecomment-940731549)). 

I tested excluding using absolute path and it did work. So it seems that if Pylance sends absolute paths, it should work too, except that it doesn't. 
The reason I'm asking this is may be if Pylance did send absolute paths, we can create a correct issue on vscode repo such as absolute path exclude outside of workspace doesn't work. IMO, that can be fixed faster than waiting for glob pattern exclude to be supported.

**Comment by heejaechang** [comment]
@binh-vu this is how we tell vscode (host) to watch those folders - https://github.com/microsoft/pyright/blob/main/packages/pyright-internal/src/languageServerBase.ts#L652

we do not work with vscode API directly but we access it through LSP (language server protocol). we do send them as absolute path but asking it to watch it recursively.

and absolute path (the workaround) they are talking about is for ```file.watcherExclude```, not for file watch itself, and the ```file.watcherExclude``` is vscode's option, not us.

I don't think it is something we should fix in our side, it must be fixed by LSP host (vscode) and fix the root cause rather than us trying to workaround or hack around it. especially since vscode is not the only host we support. 

that being said, we do support a hidden option that disable whole library file watcher itself. if you add this ```"python.analysis.watchForLibraryChanges": false``` in your ```settings.json```, we won't ask vscode to watch library folders.

currently, there is a bug where you need to reload vscode to make that option to work, but it should work as other options in next release.

**Comment by maciejmatczak** [comment]
@heejaechang, that's great to know we have a workaround! I will test it.

Currently the only workaround known to us was... uninstalling Pylance and switching back to Jedi. I was the one trying to find the root cause, as I really started to depend on Pylance ;)

**Comment by binh-vu** [comment]
@heejaechang Thanks for sharing the option. It's very nice to have a work around while waiting for the bug/new feature to be fixed/supported. 

Also, thanks for a detailed explanation. I agree that this should be fixed by the vscode team (that's why I mentioned earlier that I want to create an issue on the vscode repository)

**Comment by heejaechang** [comment]
This issue has been fixed in version 2022.7.31, which we've just released. You can find the changelog here: [CHANGELOG.md](https://github.com/microsoft/pylance-release/blob/main/CHANGELOG.md#2022731-13-july-2022)

**Comment by ethanabrooks** [comment]
@maciejmatczak did this workaround work for you? I am still receiving the same error.

**Comment by maciejmatczak** [comment]
@ethanabrooks, I believe so, I didn't see it for some time already. You should be able to narrow down which path yields the limit depletion...

**Comment by ethanabrooks** [comment]
I have an unusual setup -- I use [nix](https://nixos.org/) on my remote machine. I was also using [poetry](https://python-poetry.org/) to manage my environment. I had created a poetry environment in addition to my nix shell. I solved the issue by deleting the poetry environment and using the nix python interpreter as my vs code interpreter.
